### ISSUE TITLE
add a new button in AE2 interface to allow pushing kept stock into nearby inventories

### ISSUE THREAD

**Issue by boubou19** [GTNewHorizons/GT-New-Horizons-Modpack#9421]
### Your GTNH Discord Username

boubou_19#2706

### Your Pack Version

2.1.2.0

### Your Proposal

add a new button in AE2 interface to allow pushing kept stock into nearby inventories

### Your Goal

reduce the spam of AE2 interface + GT conveyor (+ quantium pipe) combos. Conveyors are oftenly causing lag due to their speed/checking on large inventories. AE2 pushing the kept stock it can in a single tick should reduce lag imo.

### Your Vision

interfaces already do stock keeping and have the ability to push stuff in nearby inventories when used via pattern crafting mode. This request makes it more complete.

### Final Checklist

- [X] I have searched this issue tracker and there is nothing similar already. Posting on a closed issue saying *I like this feature please reconsider adding it* will prompt us to investigate and reopen it once we confirm your report.
- [X] I believe there is nothing similar in the pack already, or the existing solution isn't good enough.
- [X] I understand this change request may not attract enough attention and thus not be implemented.
- [X] I understand this change request may be rejected due to other community members thinking it's inappropriate.
- [X] I believe this feature would make the pack better.

**Comment by bombcar** [comment]
Anything to reduce conveyor/quantum pipe spam

**Comment by Savallator** [comment]
so... like an export bus?

**Comment by boubou19** [comment]
Yeah. Endgame players just do insane I/O with the interfaces anyways. But i can understand that it might be too OP, so maybe make a card that unlocks this export function?

**Comment by Savallator** [comment]
Imho what we really need is a ME input bus just like we have the output bus.

**Comment by boubou19** [comment]
Gameplay wise, it's more interesting to do that on the interface because you can mix the two modes with one block, so cleaner setups. But yeah not gonna spit on a buff for the export bus, just that for the usages i have, it should be able to keep up with 8 stack/tick of I/O.
I'm really sick of spamming the conveyors to get speed.

What if the ME input bus would just be a regular interface but with support for GT?

**Comment by GTNH-Afx237v7** [comment]
See also #9274

**Comment by combusterf** [comment]
An ME input bus doesn't tackle even half the interface/conveyor setups I have:
- ME into chest buffer into centrifuge/electrolyser/autoclave/etc with input size requirements. I have some 50/50 between singleblocks and multis. An input bus can therefore not be used directly, export bus can't use expansion cards for more than 1 kind and retain sufficient speed.
- ME into GT pipes for preferential routing. Can't use an export bus since GT pipes refuse to connect to those.

So I prefer this suggestion, though I'd be equally happy if the above issues were solved with export buses.

**Comment by repo-alt** [comment]
I'll see if it is techinically possible for the export bus to accelerate to 1 tick per operation

**Comment by boubou19** [comment]
If it's possible to buff the export buses then could you buff the import buses as well?
because this isn't good looking:
![2022-01-09_13 18 58](https://user-images.githubusercontent.com/12850933/148681875-0edff37d-e8ad-4e76-8d61-97fe04d47dd4.png)

**Comment by repo-alt** [comment]
Looks like it is possible. I mean, it is possible right now, by setting `ExportBus.min=1` in config, but that would hit servers really hard, so there should be a separate card (with some expensive enough recipe) that would artificially remove tick limit

**Comment by boubou19** [comment]
i think the card should be gated around LuV to match the rest of the endgame AE

**Comment by Savallator** [comment]
Yeah, another problem with input  buses is that when you insert capacity cards for more filter slots it looses speed, that would need to be solved, maybe by either adding card slots , or make it use less accelleration cards for max speed.

**Comment by Savallator** [comment]
Yeah, just let it require a LuV Piston or Robot Arm or something...

**Comment by repo-alt** [comment]
This card sould probably act as top tier acceleration card in addition to the tick limit. Also LuV is maybe too early?

**Comment by boubou19** [comment]
LuV conveyors are 2 stacks/ ticks so it becomes superior to use a conveyor + interface combo

**Comment by Savallator** [comment]
Even with LuV conveyors the interface will limit the speed already, at least that is what i found. The 9 slots do not refresh every tick.

**Comment by boubou19** [comment]
no but with a LuV conveyor the interface is keeping up for 4 ticks of item movement. Also nothing stops you from putting several interfaces to absorb the massive item input. The interfaces keep up perfectly when conveyor extracts tho (if it doesn't, then your AE doesn't receive the base items to export fast enough)

**Comment by Savallator** [comment]
Then just make it LuV and nerf interface speed, but only once the new acceleration card is actually able to replace the interface in these setups

**Comment by Savallator** [comment]
But to answer the original post: It is by no means certain that this setup would actually be less laggy than conveyor+interface, because that entierely depends on how it is done. Just because it is one item vs two does not means it automatically lags less.

**Comment by boubou19** [comment]
don't touch the interface speed at all. You are going to receive a hate wave from the endgame players, me being the first. Nerfing faster ways to justify a new one is shitty. GTNH is all about having alternatives. I'm seeking for a less laggy one. I still think the interface should be able to autoexport its stock keeping mode via a button or a new card, as it'll be super easy to upgrade existing setups without doing too much changes. Swapping interfaces to export buses and reconfigure it dozen of times vs you put a card and you crowbar your conveyor.

>But to answer the original post: It is by no means certain that this setup would actually be less laggy than conveyor+interface, because that entierely depends on how it is done. Just because it is one item vs two does not means it automatically lags less.

Well GT conveyors are dumb and lose too much time checking empty slots due to their speed (hello lag on huge inventories)
Giving the export role to AE instead of GT will reduce the lag for sure as it pushes only what it can, where GT checks everything.
So on the paper it sounds less laggy.

**Comment by Savallator** [comment]
It is not about nerfing faster ways to justify a new way, it is about nerfing laggier ways to jutify a less laggy. And the argument "don't break my fancy base" is stupid in my opinion. With that mindset, we would not have Platline, Naquadah line, CAL etc...
Also, a minor speed nerf on the interface would not mean you need to swap them all immediately, but we NEED an incentive to not use that setup any more if there is an alternative, because it is laggy as you said.
And why add more functions to the already overused interface just because you used it already in many places, when there is a different item specifically for this already in the game?

**Comment by repo-alt** [comment]
In 1.18 interface is separated to pattern provider (for autocrafting) and usual iinterface (for subnetworks, etc). That is a good idea, but it is a breaking change unfortunately.

**Comment by boubou19** [comment]
>It is not about nerfing faster ways to justify a new way, it is about nerfing laggier ways to jutify a less laggy. And the argument "don't break my fancy base" is stupid in my opinion. 

It's been since more than 5 years than my base breaks +/- severly with each update. So yeah i'm used to it. The point is that to supposedly gain some lag reduction, you want to nerf something that is known to work and is the top tier in the meta to do stuff without lagging too much if used correctly. The ME output bus was made for the exact same purpose, turns out when you want to get insane speed, old combo of the output bus + conveyor lags less. But it works better for regular usages. My point is that both have their use cases. So why not keep the same speed for the interface? I'm 99% sure that both fast export bus and fast interfaces have their use case and it would be dumb to obliterate this fine balance because a specific usecase just kill the server.

Also players who don't like changing stuff won't change it until you overnerf it. We saw it with the drawer nerf. A guy even spammed the thing just to compensate the loss in #loading-screen-art. So you gain no benefit/end up getting worse because we tried to go against the player's will. GTNH has a long history with devs trying to force specific ways of doing stuff on the players (the biggest one was when Bart tried to force a specific structure on the oil crackers )

>And why add more functions to the already overused interface just because you used it already in many places, when there is a different item specifically for this already in the game?

given an interface can do both in and out, you can save channels and reduce the amount of devices on your network by building smartly. Nerfing stuff in the interface will just incentivize the players to build without trying to be smart, which will end up causing lag, thing we are trying to avoid in the first place with that change. The interface is sitting on top of the meta for those who know how to use it, and it's too handy to keep using that that almost no one will try it and will keep the interfaces + conveyors.

Finally, as @combusterf pointed out right now, only the interface is connecting to the GT pipes, not the buses, so this is yet another reason to put it on the interfaces (dunno if this is wanted/ is a bug tho)

**Comment by repo-alt** [comment]
If export bus doesn't export to a GT pipe then it is a bug (the pipe should be open at that side).
I really don't like the idea of adding yet another function to the interface (as I said I would actually like to split its functions), so I would need some arguments. (And conserving channels is more of a counter argument tbh)

**Comment by boubou19** [comment]
Well i'm not against the im/export buses to be buffed, as long as no one touch the interface in that case.

**Comment by Savallator** [comment]
I heavily use interface+conveyors as well, because there is no other way for most things. But i would welcome an interface nerf, if there is a better alternative... Existing setups should never be justification for keeping stupid mechanics.

**Comment by boubou19** [comment]
the interface is limited at 9 stacks /tick in I/O. Nerfing it will just lead people to abuse it because it'll be cheaper than the card but whatever.

**Comment by Savallator** [comment]
"I am gonna build all my automation with tin item pipes, i mean, they are cheap, yeah? And for speed i can just build more"
Also, seriously, at the point where you need the througput of several stacks per tick, some LuV parts are really free...

**Comment by boubou19** [comment]
hello SFM

**Comment by Savallator** [comment]
For nerfing SFM please make a seperate Issue...

**Comment by boubou19** [comment]
i don't ask for a nerf, i'm just finding an alternative and cheap way to bypass random pointless nerfs. I'm pretty good at it, as i must then ban them on the officials.

**Comment by Savallator** [comment]
Limiting interface speed is not a nerf, it is a performance enhancement. Did you even read what the issue is about?
I mean, i don't play on officials, but i guess even there people would like some tps sometimes, and this whole issue was about increasing performance.
And i did not even say "disable interface export", i just said to decrease speed just a tiny bit, and you defend it as if your life depends on it lol...

**Comment by boubou19** [comment]
> Did you even read what the issue is about?

yo dude that's my ticket :kekw:

> Limiting interface speed is not a nerf, it is a performance enhancement.

call it like you want, it's a regression gameplay wise for something players shouldn't have to care about: TPS. Not player's fault if something is laggy. That's the player's fault if he abuses it despite knowing it's laggy. As we both know, if perfectly used, it doesn't lag much.

> And i did not even say "disable interface export", i just said to decrease speed just a tiny bit, and you defend it as if your life depends on it lol...

That's what i tend to do because i'm fairly sure i'll end up lagging more by using those in a specific way, just like the ME output buses did in a specific way. Knowing i won't be able to come back to the previous system because of a new limitation will just worsen the TPS for nothing.

**Comment by Savallator** [comment]
I guess 90% of your conveyor+interface setups do not even need the speed of several stacks per tick anyways...

**Comment by leagris** [comment]
How about fixing the broken recipes balance that leads to these abusive amounts and rates of input and output items, that players tries to deal with, and that are impacting all players because high frequency logistics has a real material cost with CPU load ticking on servers?

**Comment by Savallator** [comment]
That would start with stuff like adding an internal filter to the VM to only actually output the ores you want (instead of 90% magnetite), include nerfing stupid recipes for bedrockium needing millions of cobble etc...

**Comment by boubou19** [comment]
and this is way out of scope of the ticket. @repo-alt i beg you: if you add the card, don't nerf the speed of the interface (not even sure it's possible) until it's extensively tested to show it's better in every way. If it doesn't, then the interface should keep its current speed.

**Comment by Savallator** [comment]
"until it's extensively tested to show it's better in every way"
That is exactly what i said as well...

**Comment by GTNH-Colen** [comment]
I think nerfing interfaces is one of the stupidest things I've ever read on the github. Do you realise the scope that they are used? I have 400+ in my base alone, this would lead to 10s of thousands of interfaces across everyones bases getting nerfed for literally no reason.

**Comment by Savallator** [comment]
This just proves the point. If you have 400+ interface.conveyor combos with high-tier conveyors this is terrible for performance, and making them just half the speed will make a huge difference.
Also, you won't have all those 400 exporting several stacks per tick, so a small nerf would not even affect most of them. What is your problem then?

**Comment by Savallator** [comment]
Ohh, nice, now people use their Github permissions to silence discussions. Enjoy then, i am out here.

**Comment by repo-alt** [comment]
If profiling doesn't show interfaces lagging much I don't see the point in nerfing them. (I also don't realy care about nerfing/buffing, I don't like one magical block doing too much, but that can't be changed unfortunately)
